### PR TITLE
feat: Support one-shot prompts

### PR DESCRIPTION
**Pull request by carlosgjs**
This PR refactors the prompt management to add support for one-shot prediction. Specifically:

- Adds the `PromptBuilder` helper class
- Changes the `prompts` module to export full prompt templates, instead of separate sets of system and instruction prompts
- Change upstream functions and tests accordingly

Test outputs from the `generate.ipynb` notebook:
```
PromptIds.AUTORA_VARS_ZEROSHOT
******* Output 0 ******** 
The dependent variable "y" is a real-valued variable, while the independent variable "x" is a continuous variable with a range of [0, 2π] and 30 allowed values. 
*************

PromptIds.AUTORA_VARS_ONESHOT
******* Output 0 ******** 
The problem is defined by one independent variable $x \in [0, 2 \pi]$ and one dependent variable $y$. 
*************
```

### CODE REVIEW

File: src/autora/doc/runtime/prompts.py [hunk 1]
```diff
@@ -1,17 +1,44 @@
...
+class PromptBuilder:
+    """
+    Utilty class for building LLAMA2 prompts. Uses a stateful builder pattern.
+    """
+
+    def __init__(self, sys: str, instr: str):
+        self.instr = instr
+        self.prompt_text = f"""<s>[INST] <<SYS>>
+{ sys }
+<</SYS>>"""
```
Comment (lsetiawan) [inline src/autora/doc/runtime/prompts.py:17]: These texts are really hard to read! I wonder if you can use [`textwrap.dedent`](https://docs.python.org/3/library/textwrap.html#textwrap.dedent) for these prompt texts?
```suggestion
        self.prompt_text = textwrap.dedent(f"""\
        <s>[INST] <<SYS>>
        { sys }
        <</SYS>>
        """)
```

File: src/autora/doc/runtime/prompts.py [hunk 2]
```diff
@@ -1,17 +1,44 @@
...
+
+    def _add_input(self) -> PromptBuilder:
+        self.prompt_text += f"""
+{ self.instr }
+----------{{code}}----------
+"""
```
Comment (lsetiawan) [inline src/autora/doc/runtime/prompts.py:23]: textwrap dedent perhaps?

File: src/autora/doc/runtime/prompts.py [hunk 3]
```diff
@@ -1,17 +1,44 @@
...
+    def add_example(self, code: str, doc: str) -> PromptBuilder:
+        self._add_input()
+        self.prompt_text = self.prompt_text.format(code=code)
+        self.prompt_text += f"""
 [/INST]
+{doc}
+</s>
+<s>
+[INST]
 """
```
Comment (lsetiawan) [inline src/autora/doc/runtime/prompts.py:35]: maybe another case of textwrap.dedent?

**Comment by lsetiawan** [comment]
The changes I think are good and I don't have much comment on that. I do have some thoughts on the `PromptBuilder` class. Is this serving almost like a template engine thing? Seems like you're trying to pack together some sort of prompt text. My thought is maybe it's worth to create some sort of prompt template say with tools like [jinja](https://jinja.palletsprojects.com/en/3.1.x/) so that it's easier to see what is actually getting build. At this moment, without actually running the code, I have a hard time understanding what is actually happening... Idk, what do you think?

File: src/autora/doc/runtime/prompts.py (hunk 1 above)
Comment (carlosgjs) [reply src/autora/doc/runtime/prompts.py]: Good idea, although I tried `textwrap.dedent` and it was very finicky about every single line having the exact trailing characters so it didn't do the trick. But I add a one-liner function to do this, along with some comments for better clarity. But I guess some familiarity with the llama2 prompt format is expected.

**Comment by carlosgjs** [comment]
> The changes I think are good and I don't have much comment on that. I do have some thoughts on the `PromptBuilder` class. Is this serving almost like a template engine thing? Seems like you're trying to pack together some sort of prompt text. My thought is maybe it's worth to create some sort of prompt template say with tools like [jinja](https://jinja.palletsprojects.com/en/3.1.x/) so that it's easier to see what is actually getting build. At this moment, without actually running the code, I have a hard time understanding what is actually happening... Idk, what do you think?

Didn't know about `jinja` and it looks pretty powerful. But perhaps its overkill for this scenario? I'm hesitant to add a package dependency for a relatively simple string building.

**Comment by lsetiawan** [comment]
> Didn't know about jinja and it looks pretty powerful. But perhaps its overkill for this scenario? I'm hesitant to add a package dependency for a relatively simple string building.

Okay, I wasn't sure how complex these prompts may be. 😄 I do like what you did in your last commit to at least make the code more readable.

**Review by lsetiawan — APPROVED** [review]

**Pull request merged**